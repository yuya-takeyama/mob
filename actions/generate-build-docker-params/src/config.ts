import { z } from 'zod';
import { readFileSync } from 'node:fs';
import { load } from 'js-yaml';
import { globSync } from 'glob';
import { dirname, join, normalize } from 'path';

const TaggingStrategySchema = z.enum([
  'always_latest',
  'semver_datetime',
  'pull_request',
]);

export type TaggingStrategy = z.infer<typeof TaggingStrategySchema>;

const DockerBuildConfigSchema = z.object({
  tagging: TaggingStrategySchema,
  registry: z.string(),
  platforms: z.array(z.string()),
});

const GitHubFilterPatternsSchema = z.union([z.string(), z.array(z.string())]);

export type GitHubFilterPatterns = z.infer<typeof GitHubFilterPatternsSchema>;

const PushEventSchema = z.object({
  branches: GitHubFilterPatternsSchema.optional(),
  branches_ignore: GitHubFilterPatternsSchema.optional(),
  tags: GitHubFilterPatternsSchema.optional(),
  tags_ignore: GitHubFilterPatternsSchema.optional(),
});

export type PushEvent = z.infer<typeof PushEventSchema>;

const PullRequestEventSchema = z.object({
  branches: GitHubFilterPatternsSchema.optional(),
  branches_ignore: GitHubFilterPatternsSchema.optional(),
});

export type PullRequestEvent = z.infer<typeof PullRequestEventSchema>;

const TriggerRuleSchema = z.object({
  push: z.union([z.literal(null), z.boolean(), PushEventSchema]).optional(),
  pull_request: z
    .union([z.literal(null), z.boolean(), PullRequestEventSchema])
    .optional(),
  pull_request_target: z
    .union([z.literal(null), z.boolean(), PullRequestEventSchema])
    .optional(),
});

export type TriggerRule = z.infer<typeof TriggerRuleSchema>;

const BuildConfigSchema = z.object({
  on: TriggerRuleSchema,
  docker: DockerBuildConfigSchema.optional(),
});

export const GlobalConfigSchema = z.object({
  build: z.object({
    docker: z.object({
      identities: z.record(
        z.string(),
        z.object({
          aws: z.object({
            'iam-role': z.string(),
            region: z.string(),
          }),
        }),
      ),
      registries: z.record(
        z.string(),
        z.object({
          aws: z.object({
            identity: z.string(),
            type: z.enum(['private', 'public']).default('private').optional(),
            region: z.string(),
            'repository-base': z.string(),
          }),
        }),
      ),
    }),
  }),
});

export type GlobalConfig = z.infer<typeof GlobalConfigSchema>;

export const LocalConfigSchema = z.object({
  build: z.array(BuildConfigSchema),
});

export type LocalConfig = z.infer<typeof LocalConfigSchema> & {
  appName: string;
  path: string;
};

export function loadGlobalConfig(filePath: string): GlobalConfig {
  const data = load(readFileSync(filePath, 'utf8'));
  return GlobalConfigSchema.parse(data);
}

export function loadLocalConfigs(rootDir: string): LocalConfig[] {
  const files = globSync('**/modjo.yaml', { cwd: rootDir });

  return files.map(file => {
    const appName = normalize(dirname(file));
    const path = normalize(join(rootDir, dirname(file)));
    const configFile = normalize(join(rootDir, file));
    return {
      appName,
      path,
      ...loadLocalConfig(configFile),
    };
  });
}

function loadLocalConfig(filePath: string) {
  const data = load(readFileSync(filePath, 'utf8'));
  return LocalConfigSchema.parse(data);
}
