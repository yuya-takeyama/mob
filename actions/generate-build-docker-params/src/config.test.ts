import { loadGlobalConfig, loadLocalConfigs } from './config';

describe('loadloadGlobalConfig', () => {
  it('should load valid config file', () => {
    const config = loadGlobalConfig('fixtures/apps/modjo-global.yaml');
    expect(config).toMatchSnapshot();
  });
});

describe('loadLocalConfigs', () => {
  it('should load valid config files', () => {
    const configs = loadLocalConfigs('fixtures/apps');
    expect(configs).toMatchSnapshot();
  });
});
