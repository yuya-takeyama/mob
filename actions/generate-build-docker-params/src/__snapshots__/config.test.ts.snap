// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadLocalConfigs should load valid config files 1`] = `
[
  {
    "appName": "hello-world",
    "build": [
      {
        "docker": {
          "platforms": [
            "linux/amd64",
          ],
          "registry": "stg-pr",
          "tagging": "pull_request",
        },
        "on": {
          "pull_request": null,
        },
      },
      {
        "docker": {
          "platforms": [
            "linux/amd64",
          ],
          "registry": "stg-main",
          "tagging": "always_latest",
        },
        "on": {
          "push": {
            "branches": [
              "main",
            ],
          },
        },
      },
      {
        "docker": {
          "platforms": [
            "linux/amd64",
          ],
          "registry": "prd",
          "tagging": "semver_datetime",
        },
        "on": {
          "push": {
            "branches": [
              "main",
            ],
          },
        },
      },
    ],
    "path": "fixtures/apps/hello-world",
  },
]
`;

exports[`loadloadGlobalConfig should load valid config file 1`] = `
{
  "build": {
    "docker": {
      "identities": {
        "prd": {
          "aws": {
            "iam-role": "arn:aws:iam::920373013500:role/mobb-builder",
            "region": "ap-northeast-1",
          },
        },
        "stg-main": {
          "aws": {
            "iam-role": "arn:aws:iam::615299752259:role/mobb-builder-main",
            "region": "ap-northeast-1",
          },
        },
        "stg-pr": {
          "aws": {
            "iam-role": "arn:aws:iam::615299752259:role/mobb-builder-pr",
            "region": "ap-northeast-1",
          },
        },
      },
      "registries": {
        "prd": {
          "aws": {
            "identity": "prd",
            "region": "ap-northeast-1",
            "repository-base": "920373013500.dkr.ecr.ap-northeast-1.amazonaws.com/mobb",
          },
        },
        "stg-main": {
          "aws": {
            "identity": "stg-main",
            "region": "ap-northeast-1",
            "repository-base": "615299752259.dkr.ecr.ap-northeast-1.amazonaws.com/mobb",
          },
        },
        "stg-pr": {
          "aws": {
            "identity": "stg-pr",
            "region": "ap-northeast-1",
            "repository-base": "615299752259.dkr.ecr.ap-northeast-1.amazonaws.com/mobb-pr",
          },
        },
      },
    },
  },
}
`;
